File: algorithm/003-longest_substring_without_repeating_characters/java/solution.java
class Solution {
  public int lengthOfLongestSubstring(String s) {
    int[] map = new int[127];
    int longest = 0;
    int i = 0; // 滑窗左边界
    int j = 0; // 滑窗右边界

    for ( ; j < s.length(); j++) {
      char c = s.charAt(j);
      if (map[c] > 0) {
        longest = Math.max(longest, j - i);
        while (i < map[c.toInt()]) {
          map[s[i].toInt()] = 0
          i++
        }
      }

      map[c] = j + 1;
    }

    return Math.max(longest, j - i);
  }
}

File: algorithm/003-longest_substring_without_repeating_characters/scala/solution.scala
object Solution {
  def lengthOfLongestSubstring(s: String): Int = {
    val alphabet = new Array[Int](127)
    var longest = 0
    var i = 0
    var j = 0

    s.foreach( c => {
      if (alphabet(c) > 0) {
        longest = math.max(longest, j - i)
        while (i < alphabet(c)) {
          alphabet(s(i)) = 0
          i += 1
        }
      }

      j += 1
      alphabet(c) = j
    })

    return math.max(longest, j - i)
  }
}